File: tools/testing/connection/undici/rally.mjs
/**
 * Rally API - undici Connection Test
 *
 * Note: Rally integration is currently a placeholder and not yet implemented.
 *
 * Authentication: ZSESSIONID or API Key
 * Base URL: https://rally1.rallydev.com/slm/webservice/v2.0
 * Health Endpoint: GET /subscription
 *
 * TLS/SSL Options:
 *   NODE_TLS_REJECT_UNAUTHORIZED=0  - Ignore all certificate errors
 *   REQUEST_CA_BUNDLE=/path/to/ca   - Custom CA bundle file
 *   SSL_CERT_FILE=/path/to/cert     - Custom SSL certificate file
 *   NODE_EXTRA_CA_CERTS=/path/to/ca - Additional CA certificates
 */

import { request, ProxyAgent, Agent } from 'undici';
import { readFileSync } from 'node:fs';

// ============================================================================
// Configuration - Override these values
// ============================================================================

const CONFIG = {
  // Required
  RALLY_API_KEY: process.env.RALLY_API_KEY || 'your_rally_api_key_here',

  // Base URL
  BASE_URL: process.env.RALLY_BASE_URL || 'https://rally1.rallydev.com/slm/webservice/v2.0',

  // Optional: Proxy Configuration
  HTTPS_PROXY: process.env.HTTPS_PROXY || '', // e.g., 'http://proxy.example.com:8080'

  // Optional: TLS Configuration
  // Set NODE_TLS_REJECT_UNAUTHORIZED=0 to ignore certificate errors
  REJECT_UNAUTHORIZED: process.env.NODE_TLS_REJECT_UNAUTHORIZED !== '0',

  // Optional: Custom CA certificates
  // REQUEST_CA_BUNDLE or SSL_CERT_FILE - path to custom CA bundle
  // NODE_EXTRA_CA_CERTS - handled automatically by Node.js
  CA_BUNDLE: process.env.REQUEST_CA_BUNDLE || process.env.SSL_CERT_FILE || '',
};

// ============================================================================
// TLS Configuration Helper
// ============================================================================

function getTlsConnectOptions() {
  const options = {
    rejectUnauthorized: CONFIG.REJECT_UNAUTHORIZED,
  };

  // Load custom CA bundle if specified
  if (CONFIG.CA_BUNDLE) {
    try {
      options.ca = readFileSync(CONFIG.CA_BUNDLE);
      console.log(`Using custom CA bundle: ${CONFIG.CA_BUNDLE}`);
    } catch (err) {
      console.warn(`Warning: Could not read CA bundle: ${err.message}`);
    }
  }

  return options;
}

// ============================================================================
// Create Dispatcher (with or without proxy)
// ============================================================================

function createDispatcher() {
  const tlsOptions = getTlsConnectOptions();

  if (CONFIG.HTTPS_PROXY) {
    console.log(`Using proxy: ${CONFIG.HTTPS_PROXY}`);
    return new ProxyAgent({
      uri: CONFIG.HTTPS_PROXY,
      connect: tlsOptions,
    });
  }
  return new Agent({
    connect: tlsOptions,
  });
}

// ============================================================================
// Health Check
// ============================================================================

async function healthCheck() {
  console.log('\n=== Rally Health Check ===\n');
  console.log('Note: Rally integration is a placeholder - not yet implemented.\n');

  const dispatcher = createDispatcher();

  try {
    const response = await request(`${CONFIG.BASE_URL}/subscription`, {
      method: 'GET',
      headers: {
        'ZSESSIONID': CONFIG.RALLY_API_KEY,
        'Content-Type': 'application/json',
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log('Response:', JSON.stringify(data, null, 2));

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

// ============================================================================
// Sample API Calls (Reference)
// ============================================================================

async function getCurrentUser() {
  console.log('\n=== Get Current User ===\n');

  const dispatcher = createDispatcher();

  try {
    const response = await request(`${CONFIG.BASE_URL}/user`, {
      method: 'GET',
      headers: {
        'ZSESSIONID': CONFIG.RALLY_API_KEY,
        'Content-Type': 'application/json',
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log('Response:', JSON.stringify(data, null, 2));

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

async function listProjects() {
  console.log('\n=== List Projects ===\n');

  const dispatcher = createDispatcher();

  try {
    const response = await request(`${CONFIG.BASE_URL}/project`, {
      method: 'GET',
      headers: {
        'ZSESSIONID': CONFIG.RALLY_API_KEY,
        'Content-Type': 'application/json',
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    const results = data.QueryResult?.Results || [];
    console.log(`Found ${results.length} projects`);
    results.slice(0, 10).forEach((project) => {
      console.log(`  - ${project.Name}`);
    });

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

async function queryUserStories(projectName) {
  console.log(`\n=== Query User Stories: ${projectName} ===\n`);

  const dispatcher = createDispatcher();

  try {
    const query = encodeURIComponent(`(Project.Name = "${projectName}")`);
    const response = await request(`${CONFIG.BASE_URL}/hierarchicalrequirement?query=${query}`, {
      method: 'GET',
      headers: {
        'ZSESSIONID': CONFIG.RALLY_API_KEY,
        'Content-Type': 'application/json',
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    const results = data.QueryResult?.Results || [];
    console.log(`Found ${results.length} user stories`);
    results.slice(0, 10).forEach((story) => {
      console.log(`  - ${story.FormattedID}: ${story.Name}`);
    });

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

async function listDefects() {
  console.log('\n=== List Defects ===\n');

  const dispatcher = createDispatcher();

  try {
    const response = await request(`${CONFIG.BASE_URL}/defect`, {
      method: 'GET',
      headers: {
        'ZSESSIONID': CONFIG.RALLY_API_KEY,
        'Content-Type': 'application/json',
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    const results = data.QueryResult?.Results || [];
    console.log(`Found ${results.length} defects`);
    results.slice(0, 10).forEach((defect) => {
      console.log(`  - ${defect.FormattedID}: ${defect.Name}`);
    });

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

// ============================================================================
// Run Tests
// ============================================================================

async function main() {
  console.log('Rally API Connection Test');
  console.log('=========================');
  console.log('Status: PLACEHOLDER - Not yet implemented\n');
  console.log(`Base URL: ${CONFIG.BASE_URL}`);
  console.log(`Proxy: ${CONFIG.HTTPS_PROXY || 'None'}`);
  console.log(`API Key: ${CONFIG.RALLY_API_KEY.slice(0, 10)}...`);
  console.log(`TLS Verify: ${CONFIG.REJECT_UNAUTHORIZED}`);
  console.log(`CA Bundle: ${CONFIG.CA_BUNDLE || 'System default'}`);

  await healthCheck();
  // await getCurrentUser();
  // await listProjects();
  // await queryUserStories('My Project');
  // await listDefects();
}

main().catch(console.error);

File: tools/testing/connection/httpx/rally.py
#!/usr/bin/env python3
"""
Rally API - httpx Connection Test

Authentication: ZSESSIONID header or API Key
Base URL: https://rally1.rallydev.com/slm/webservice/v2.0
Health Endpoint: GET /security/authorize

TLS/SSL Options:
  SSL_CERT_VERIFY=0              - Ignore all certificate errors
  REQUEST_CA_BUNDLE=/path/to/ca  - Custom CA bundle file
  SSL_CERT_FILE=/path/to/cert    - Custom SSL certificate file
  REQUESTS_CA_BUNDLE=/path/to/ca - Alternative CA bundle (requests compat)
"""

import asyncio
import json
import os
from typing import Any, Union

import httpx

# ============================================================================
# Configuration - Override these values
# ============================================================================

CONFIG = {
    # Required (use one of these methods)
    "RALLY_API_KEY": os.getenv("RALLY_API_KEY", ""),  # Preferred method
    "RALLY_ZSESSIONID": os.getenv("RALLY_ZSESSIONID", ""),  # Alternative

    # Base URL
    "BASE_URL": os.getenv("RALLY_BASE_URL", "https://rally1.rallydev.com/slm/webservice/v2.0"),

    # Optional: Workspace
    "RALLY_WORKSPACE": os.getenv("RALLY_WORKSPACE", ""),

    # Optional: Proxy Configuration
    "HTTPS_PROXY": os.getenv("HTTPS_PROXY", ""),  # e.g., "http://proxy.example.com:8080"

    # Optional: TLS Configuration
    # Set SSL_CERT_VERIFY=0 to ignore certificate errors
    "VERIFY_SSL": os.getenv("SSL_CERT_VERIFY", "1") != "0",

    # Optional: Custom CA certificates
    # REQUEST_CA_BUNDLE, SSL_CERT_FILE, or REQUESTS_CA_BUNDLE - path to custom CA bundle
    "CA_BUNDLE": os.getenv("REQUEST_CA_BUNDLE") or os.getenv("SSL_CERT_FILE") or os.getenv("REQUESTS_CA_BUNDLE") or "",
}


# ============================================================================
# TLS Configuration Helper
# ============================================================================

def get_ssl_verify() -> Union[bool, str]:
    """Get SSL verification setting."""
    if CONFIG["CA_BUNDLE"]:
        print(f"Using custom CA bundle: {CONFIG['CA_BUNDLE']}")
        return CONFIG["CA_BUNDLE"]
    return CONFIG["VERIFY_SSL"]


# ============================================================================
# Create HTTP Client
# ============================================================================

def create_client() -> httpx.AsyncClient:
    """Create httpx async client with optional proxy."""
    proxies = None
    if CONFIG["HTTPS_PROXY"]:
        print(f"Using proxy: {CONFIG['HTTPS_PROXY']}")
        proxies = {
            "http://": CONFIG["HTTPS_PROXY"],
            "https://": CONFIG["HTTPS_PROXY"],
        }

    return httpx.AsyncClient(
        proxies=proxies,
        verify=get_ssl_verify(),
        timeout=httpx.Timeout(30.0),
    )


def create_auth_headers() -> dict[str, str]:
    """Create authentication headers."""
    headers = {
        "Accept": "application/json",
        "Content-Type": "application/json",
    }

    if CONFIG["RALLY_API_KEY"]:
        headers["ZSESSIONID"] = CONFIG["RALLY_API_KEY"]
    elif CONFIG["RALLY_ZSESSIONID"]:
        headers["ZSESSIONID"] = CONFIG["RALLY_ZSESSIONID"]

    return headers


# ============================================================================
# Health Check
# ============================================================================

async def health_check() -> dict[str, Any]:
    """Perform health check against Rally API."""
    print("\n=== Rally Health Check ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{CONFIG['BASE_URL']}/security/authorize",
                headers=create_auth_headers(),
            )

            data = response.json()

            print(f"Status: {response.status_code}")

            operation_result = data.get("OperationResult", {})
            if operation_result.get("SecurityToken"):
                print(f"Security Token: {operation_result['SecurityToken'][:20]}...")
                print("Authentication: Success")
            else:
                errors = operation_result.get("Errors", [])
                if errors:
                    print(f"Errors: {errors}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


# ============================================================================
# Sample API Calls
# ============================================================================

async def get_user() -> dict[str, Any]:
    """Get current user information."""
    print("\n=== Get Current User ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{CONFIG['BASE_URL']}/user",
                headers=create_auth_headers(),
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            user = data.get("User", {})
            print(f"User: {user.get('_refObjectName')}")
            print(f"Email: {user.get('EmailAddress')}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


async def list_workspaces() -> dict[str, Any]:
    """List available workspaces."""
    print("\n=== List Workspaces ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{CONFIG['BASE_URL']}/subscription",
                params={"fetch": "Workspaces"},
                headers=create_auth_headers(),
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            subscription = data.get("Subscription", {})
            workspaces = subscription.get("Workspaces", {}).get("_tagsNameArray", [])
            print(f"Found {len(workspaces)} workspaces")
            for ws in workspaces[:10]:
                print(f"  - {ws.get('Name')}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


async def list_projects(workspace_ref: str = "") -> dict[str, Any]:
    """List projects in a workspace."""
    print("\n=== List Projects ===\n")

    async with create_client() as client:
        try:
            params = {"fetch": "Name,ObjectID", "pagesize": 20}
            if workspace_ref:
                params["workspace"] = workspace_ref

            response = await client.get(
                f"{CONFIG['BASE_URL']}/project",
                params=params,
                headers=create_auth_headers(),
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            query_result = data.get("QueryResult", {})
            results = query_result.get("Results", [])
            print(f"Found {query_result.get('TotalResultCount', 0)} projects")
            for project in results[:10]:
                print(f"  - {project.get('Name')} ({project.get('ObjectID')})")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


async def query_user_stories(project_ref: str = "", query: str = "") -> dict[str, Any]:
    """Query user stories."""
    print("\n=== Query User Stories ===\n")

    async with create_client() as client:
        try:
            params = {
                "fetch": "FormattedID,Name,ScheduleState,Owner",
                "pagesize": 10,
                "order": "FormattedID desc",
            }
            if project_ref:
                params["project"] = project_ref
            if query:
                params["query"] = query

            response = await client.get(
                f"{CONFIG['BASE_URL']}/hierarchicalrequirement",
                params=params,
                headers=create_auth_headers(),
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            query_result = data.get("QueryResult", {})
            results = query_result.get("Results", [])
            print(f"Found {query_result.get('TotalResultCount', 0)} stories")
            for story in results[:10]:
                owner = story.get("Owner", {})
                owner_name = owner.get("_refObjectName", "Unassigned") if owner else "Unassigned"
                print(f"  - {story.get('FormattedID')}: {story.get('Name')} ({story.get('ScheduleState')}) - {owner_name}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


async def query_defects(project_ref: str = "", query: str = "") -> dict[str, Any]:
    """Query defects."""
    print("\n=== Query Defects ===\n")

    async with create_client() as client:
        try:
            params = {
                "fetch": "FormattedID,Name,State,Priority,Severity",
                "pagesize": 10,
                "order": "FormattedID desc",
            }
            if project_ref:
                params["project"] = project_ref
            if query:
                params["query"] = query

            response = await client.get(
                f"{CONFIG['BASE_URL']}/defect",
                params=params,
                headers=create_auth_headers(),
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            query_result = data.get("QueryResult", {})
            results = query_result.get("Results", [])
            print(f"Found {query_result.get('TotalResultCount', 0)} defects")
            for defect in results[:10]:
                print(f"  - {defect.get('FormattedID')}: {defect.get('Name')} ({defect.get('State')}) - {defect.get('Priority')}/{defect.get('Severity')}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


# ============================================================================
# Run Tests
# ============================================================================

async def main():
    """Run connection tests."""
    print("Rally API Connection Test")
    print("=========================")
    print(f"Base URL: {CONFIG['BASE_URL']}")
    print(f"Proxy: {CONFIG['HTTPS_PROXY'] or 'None'}")
    print(f"API Key: {'Configured' if CONFIG['RALLY_API_KEY'] else 'Not configured'}")
    print(f"ZSESSIONID: {'Configured' if CONFIG['RALLY_ZSESSIONID'] else 'Not configured'}")
    print(f"SSL Verify: {CONFIG['VERIFY_SSL']}")
    print(f"CA Bundle: {CONFIG['CA_BUNDLE'] or 'System default'}")

    await health_check()
    # await get_user()
    # await list_workspaces()
    # await list_projects()
    # await query_user_stories()
    # await query_defects()


if __name__ == "__main__":
    asyncio.run(main())
